File: 2025-03-March-GFG-POTD/March 18 - Partition Equal Subset Sum/Solution.java
//{ Driver Code Starts
// Initial Template for Java

import java.io.*;
import java.util.*;

class GFG {
    public static void main(String args[]) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());
        while (t-- > 0) {
            String inputLine[] = br.readLine().trim().split(" ");
            int n = inputLine.length;
            int arr[] = new int[n];
            for (int i = 0; i < n; i++) {
                arr[i] = Integer.parseInt(inputLine[i]);
            }

            Solution ob = new Solution();

            if (ob.equalPartition(arr))
                System.out.println("true");
            else
                System.out.println("false");

            System.out.println("~");
        }
    }
}
// } Driver Code Ends


class Solution {
    static boolean equalPartition(int arr[]) {
        int sum = 0;
        for (int num : arr) {
            sum += num;
        }

        // If total sum is odd, it's impossible to split into equal subsets
        if (sum % 2 != 0) return false;

        int target = sum / 2;
        int n = arr.length;
        boolean dp[][] = new boolean[n + 1][target + 1];

        // Base case: sum 0 is always possible
        for (int i = 0; i <= n; i++) {
            dp[i][0] = true;
        }

        // DP table filling
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= target; j++) {
                if (arr[i - 1] > j) { // If element is greater than sum, exclude it
                    dp[i][j] = dp[i - 1][j];
                } else {
                    // Include or exclude the current element
                    dp[i][j] = dp[i - 1][j] || dp[i - 1][j - arr[i - 1]];
                }
            }
        }

        return dp[n][target];
    }
}

File: 2025-03-March-GFG-POTD/March 18 - Partition Equal Subset Sum/Solution.py
class Solution:
    def equalPartition(self, arr):
        total_sum = sum(arr)

        # If total sum is odd, it's impossible to split into equal subsets
        if total_sum % 2 != 0:
            return False

        target = total_sum // 2
        n = len(arr)
        dp = [[False] * (target + 1) for _ in range(n + 1)]

        # Base case: sum 0 is always possible
        for i in range(n + 1):
            dp[i][0] = True

        # DP table filling
        for i in range(1, n + 1):
            for j in range(1, target + 1):
                if arr[i - 1] > j:  # If element is greater than sum, exclude it
                    dp[i][j] = dp[i - 1][j]
                else:
                    # Include or exclude the current element
                    dp[i][j] = dp[i - 1][j] or dp[i - 1][j - arr[i - 1]]

        return dp[n][target]


# Driver Code
if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        arr = list(map(int, input().split()))
        ob = Solution()
        print("true" if ob.equalPartition(arr) else "false")
        print("~")
